gk/main.c

```diff
@@ -717,3 +829,92 @@ cleanup_gk(struct gk_config *gk_conf)
 
 	return 0;
 }
+
+int
+get_responsible_block(struct ip_flow *flow, struct gk_config *gk_conf)
+{
+	/*
+	 * Calculate the RSS hash value for the
+	 * pair <Src, Dst> in the decision.
+	 */
+	uint32_t rss_hash_val = rss_ip_flow_hf(flow, 0, 0);
+	uint32_t i;
+	uint32_t idx;
+	uint32_t shift;
+	uint16_t queue_id;
+	int block_id = -1;
+
+	/*
+	 * XXX Change the mapping rss hash value to rss reta entry
+	 * if the reta size is not 128.
+	 */
+	RTE_ASSERT(gk_conf->rss_conf.reta_size == 128);
+	rss_hash_val = (rss_hash_val & 127);
+
+	idx = rss_hash_val / RTE_RETA_GROUP_SIZE;
+	shift = rss_hash_val % RTE_RETA_GROUP_SIZE;
+	queue_id = gk_conf->rss_conf.reta_conf[idx].reta[shift];
+
+	/* XXX Change mapping queue id to the gk instance id efficiently. */
+	for (i = gk_conf->lcore_start_id; i<= gk_conf->lcore_end_id; i++) {
+		if (gk_conf->instances[i - gk_conf->lcore_start_id].
+				rx_queue_front == queue_id)
+			block_id = i - gk_conf->lcore_start_id;
+	}
+
+	if (block_id == -1)
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: wrong RSS configuration for GK blocks!\n");
+
+	return block_id;
+}
+
+int
+send_gk_cmd(unsigned int block_idx, struct gk_cmd_entry *entry,
+	struct gk_config *gk_conf)
+{
+	int ret = -1;
+	struct gk_instance *instance;
+	struct gk_cmd_entry *cmd_entry = NULL;
+
+	if (gk_conf->lcore_start_id + block_idx > gk_conf->lcore_end_id)
+		goto out;
+
+	instance = &gk_conf->instances[block_idx];
+
+	ret = rte_mempool_get(instance->mb->pool, (void **)&cmd_entry);
```

Comment:
This call is a violation of the mailbox abstraction. The mailbox should have a function that just returns a `cmd_entry`.

Moreover, is that cast `(void **)` really needed?